[Headers][X86] Allow SLLDQ/SRLDQ byte shift intrinsics to be used in constexpr

TianYe717 opened the pull request:
Support constexpr usage for SLLDQ/SRLDQ byte shift intrinsics

This draft PR adds support for using the following SRLDQ intrinsics in constant expressions:
- _mm_srli_si128
- _mm256_srli_si256
- _mm_slli_si128
- _mm256_slli_si256
- _mm512_bslli_epi128
- _mm512_bsrli_epi128

Relevant tests are included.

Fixes #156494

TianYe717 committed:
[Headers][X86] Support constexpr usage for _mm_srli_si128 and _mm256_srli_si256 intrinsics

TianYe717 committed:
[Headers][X86] Support constexpr usage for _mm_slli_si128 and _mm256_slli_si256 intrinsics

TianYe717 committed:
[Headers][X86] Support constexpr usage for _mm_srli_si128 and _mm256_srli_si256

TianYe717 commented:
Hi @RKSimon @tbaederr , could you please take a look at this PR? Thanks!

TianYe717 committed:
[Headers][X86] Address review comments for SLLDQ/SRLDQ constexpr intrinsics

TianYe717 committed:
[NFC] Code format

RKSimon commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -12790,6 +12790,64 @@ bool VectorExprEvaluator::VisitCallExpr(const CallExpr *E) {
+      return false;
+
+    unsigned VecLen = Src.getVectorLength();
+    unsigned Shift = Imm.getZExtValue();
```
```suggestion
    unsigned Shift = Imm.getZExtValue() & 0xff;
```

RKSimon commented on clang/lib/AST/ByteCode/InterpBuiltin.cpp:
```diff
@@ -3128,6 +3128,38 @@ static bool interp__builtin_ia32_vpconflict(InterpState &S, CodePtr OpPC,
+  assert(Call->getNumArgs() == 2);
+
+  APSInt ImmAPS = popToAPSInt(S, Call->getArg(1));
+  uint64_t Shift = ImmAPS.getZExtValue();
```
Consult EmitX86BuiltinExpr - we only the bottom imm8
```suggestion
  uint64_t Shift = ImmAPS.getZExtValue() & 0xff;
```

TianYe717 committed:
[Headers][X86] Simplify SLLDQ/SRLDQ lambda: hardcode bit width for byte shift intrinsic

- Hardcode APInt width to 8 bits in SLLDQ/SRLDQ byte shift lambdas, matching the intrinsic's semantics.
- Added clarifying comments explaining lane width and element width relationship.

TianYe717 committed:
[Headers][X86] Mask shift immediate to 8 bits for byte shift intrinsic

TianYe717 committed:
[Headers][X86] Address review comments for SLLDQ/SRLDQ constexpr intrinsics

RKSimon reviewed (changes requested):
Missing _mm512_bslli_epi128 / _mm512_bsrli_epi128 test coverage in avx512bw-builtins.c

TianYe717 committed:
[Headers][X86] Support _mm512_bslli_epi128 / _mm512_bsrli_epi128 and add test coverage

TianYe717 commented:
> Missing _mm512_bslli_epi128 / _mm512_bsrli_epi128 test coverage in avx512bw-builtins.c

Added support and tests for these intrinsics, thanks!

RKSimon reviewed (approved):
LGTM - cheers

RKSimon committed:
Merge branch 'main' into x86-slldq-srldq-constexpr
